### PR TITLE
pink: Permission control for on_block_end

### PR DESCRIPTION
**Pull request by kvinwang**
This PR adds permission control for the hook on_block_end.

- Add host API `fn set_hook(hook_point,contract,selector)` where hook_point can be `on_block_end`.
- The host only accepts setting hook from the system contract.
- Add pink::System API `fn set_hook(hook_point,contract,selector)`.
- The default pink-system contract allows contracts with admin permission to set_hook.
- The macro `#[pink(on_block_end)]` was removed.Need to call set_hook explicitly now.

We can implement more fine-grained access control in a driver contract in the future.

### CODE REVIEW

File: crates/pink/pink-extension/src/lib.rs [hunk 1]
```diff
@@ -19,10 +19,15 @@ pub mod predefined_accounts {
 
     // TODO.kevin: Should move to a separate crates. Maybe after https://github.com/Phala-Network/phala-blockchain/issues/861 resolved.
     pub const ACCOUNT_PALLET: [u8; 32] = *b"sys::pellet\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0";
+    pub const ACCOUNT_RUNTIME: [u8; 32] = [0; 32];
```
Comment (h4x3rotab) [inline crates/pink/pink-extension/src/lib.rs:22]: Better to not use completely zero address. It's easy to accidentally pass a zero account, causing vulnerabilities

**Review by h4x3rotab — APPROVED** [review]
LGTM.

Btw, what will happen if `on_block_end()` fails?

File: crates/pink/pink-extension/src/lib.rs (hunk 1 above)
Comment (kvinwang) [reply crates/pink/pink-extension/src/lib.rs]: Fixed

**Comment by kvinwang** [comment]
> Btw, what will happen if on_block_end() fails?

Internally, it is a normal `ink(message)` call, So it would roll back the state changes if fails.

**Pull request merged**